File: Problem-1_reverse_linked_list.py
# APPROACH 1: ITERATIVE APPROACH
# Time Complexity : O(n), n: length of the linked list
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. Have a pointer pointing to previous, current and 
# 2. In each iteration, current element's next points to previous element
# 3. Update previous -> current, current -> next and next -> next's next

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        
        if head is None or head.next is None:
            return head
        
        prev, curr, curr_next = None, head, head.next
        
        while curr_next is not None:
            curr.next = prev
            prev = curr
            curr = curr_next
            curr_next = curr_next.next
            
        curr.next = prev
        
        return curr
        
        
        
        
        
        

# APPROACH 2: RECURSIVE APPROACH
# Time Complexity : O(n), n: length of the linked list
# Space Complexity : O(n), due to the size of the recursive stack
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None
#
#
# Your code here along with comments explaining your approach
# 1. Traverse till the end of the list
# 2. From back onwards, each node's next element's next will point to the node and make each node's next point to None (either will get updated in next recursive call or last 
#    element points to None
# 3. Make sure you keep passing the startnode (original list's last node), so that we can access the list

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        
        if head is None or head.next is None:
            return head
        
        start_node = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        
        return start_node

File: Problem1_ReverseLinkedList.cs
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
// Approach: The intution is to use 3 pointers which points to current node, its previous node and its next node,
// and update the link of current node to previous node so that the link is reversed.
 
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {
        if(head == null) return head;

        ListNode prev = null; //pointer to previous node
        ListNode temp; //pointer to next node

        while(head != null) {
            temp = head.next;
            head.next = prev; //change the link of head from next node to previous node.
            prev = head; //move the prev pointer
            head = temp; //move head pointer
        }
        return prev;
    }
}